### DISCUSSION TITLE
Victron MPPT korrekt via Ve.Direct mit dem ESP verbinden

### DISCUSSION THREAD

**Discussion by swingstate** [hoylabs/OpenDTU-OnBattery#333] [General]
Moin Community, 

ich bin mit meinem Setup (Umbau von AC auf direktes DC Laden der Batterien) fast fertig und möchte nun den MPPT via VE.Direct an den ESP hängen um den SOC auszulesen. 
Ein Kabel, welches RX und TX führt (die beiden mittleren PINs des VE.Direct Anschlusses) habe ich mir gebaut. Meine DTU habe ich bei Kleinanzeigen gekauft, leider sind die PIN Beschriftungen nicht sichtbar. Nach meiner Interpretation handelt es sich bei meinem Model aber um die "Standard" DTU Verdrahtung, daher habe ich vor die GPIOs 32 & 33 zu nutzen, Das sind nach meinem Verständnis die PINs 6 & 7.

Passt das? Das mapping mache ich dann in der Hardware config via Json upload in die Webui.

![opendtu](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/73672554/3b8a6919-9a1a-4f56-8c20-ca521e8759bc)

**Comment by madmartin** [comment]
Hallo @swingstate , na das sieht ja aus....
Also 
als erstes mußt Du zählen wieviele Pins das ESP32 Modul hat. Es gibt welche mit 30pin und 38pin.
Dann schaust Du auf das entsprechende Bild, orientierst Dich am USB-Anschluss und mußt das entsprechend im Geiste umdrehen, denn die Bilder zeigen immer die Seite mit den Bauteilen, während bei Dir das Modul sozusagen auf dem Rücken liegt.

Was Du da hast sieht mir nach einem 38pin Modul aus. Dann sind GPIO 32 33  von oben gezählt Pin 7 und 8.

![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/4208120/6948b420-4a27-4475-a24a-4060aa172208)


![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/4208120/3723bfa9-6a24-4df4-9bae-0c755acde1be)

**Comment by madmartin** [comment]
Kleine Anmerkung noch:

> und möchte nun den MPPT via VE.Direct an den ESP hängen um den SOC auszulesen. 

Der MPPT Lader kann den SoC nicht lesen. Aber die Batteriespannung wird angezeigt, und über diese lässt sich der Power Limiter dann steuern.
Der "echte" SoC wird nur vom Batterie-Management geliefert - wenn vorhanden.

**Comment by swingstate** [comment]
Danke @madmartin !

Ich war auch erheitert als ich den Kasten das erste Mal aufgemacht habe. Unabhängig von der Unordnung funkioniert die DTU seit Monaten jedoch problemlos. 

Danke für den Hinweiß zum SOC, rechte haste. Den SOC lasse ich mir von einem Victron Smartshunt berechnen und in Homeassistant via ESPHome anzeigen bzw. nutze die Werte dort für Automatisierungen. Die 2 Powerqueen Akkus liefern leider keine Daten nach außen.

**Comment by swingstate** [comment]
So, habe alles verdrahtet und mein HM-1500 wird nun durch 2x 25,6V gespeist. DPL habe ich aktiv, Ve.Direct ebenfalls konfiguriert. Woran kann ich denn sehen, dass die DTU korrekte Daten via Ve.Direct empfängt? Ich sehe auf Anhieb nichts in der Konsole.

**Reply by helgeerbe** [reply]
Live View! Ggf. in den Setting ve.direct noch aktivieren.
![Cursor_und_OpenDTU-onBattery](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/59169507/b5124032-91fa-458a-95c7-19e90a00477c)

**Comment by schlimmchen** [comment]
> Ich sehe auf Anhieb nichts in der Konsole.

@swingstate Der VE.Direct FrameHandler ist "still" -- außer bei Fehlern. Wie Helge schrieb sollte bei funktionierendem VE.Direct sofort ein neuer Block im LiveView auftauchen.

**Reply by swingstate** [reply]
Ja, den Block sehe ich, allerdings ist der Header rot und es gibt keine Daten, weder in dem unteren Block noch oben in den "MPPT" Details. Ich vermute, dass ich die PINs nicht korrekt verbunden habe. Kann ich erst wieder am Samstag troubeshooten.

**Comment by swingstate** [comment]
Ich nun Zeit weiter am VE.Direct zu arbeiten und habe dabei augenscheinlich eine fehlerhafte pin_mapping.json hochgeladen. Die DTU kommt nun nicht mehr hoch. Neustarten und neu kompilieren & flashen hilft nicht. Wie kann ich die DTU komplett zurück setzen oder lässt sich das PIN mapping anders korrigieren?

**Reply by spcqike** [reply]
Im Zweifel vor dem seriellen Flashen den Flash komplett löschen. 

Das pin mapping ist ja eine Datei im Dateisystem welches bei einem einfachen Firmware Update nicht angefasst wird.

**Comment by swingstate** [comment]
Ich habe den ESP mit den ESPtools gelöscht und via VSCode neu geflasht. Ich sehe bei den PINs in der Hardware config nun überall -1 stehen und kann keine Verbindung zum WR aufbauen. Ich meine dass schon einmal hier gelesen zu haben, kann die Lösung aber aktuell nicht finden. 

Ah ... #281 na toll.

**Reply by madmartin** [reply]
Dann nimm eine leere Device-Profiles-Datei (Unterverzeichnis `docs/DeviceProfiles`), editiere sie mit einem Texteditor, dann hochladen auf die OpenDTU-onBattery, und dann im Device-Manager das erstellte und hochgeladene Profil auswählen.

**Comment by swingstate** [comment]
Danke Martin. Verbindung zum WR steht, jedoch bekomme ich keine Verbindung zum MPPT.

Ich habe nochmals ein einfaches Kabel gebaut. An dem JST Stecker der im MPPT steckt nutze ich die zwei mittleren Adern. Diese sind auf GPIO 32 & 33 verbunden. 

Dazu habe ich zwei Profile angelegt, und die Konfiguration dieser scheint auch zu klappen.

<img width="1100" alt="Bildschirmfoto 2023-07-28 um 21 08 17" src="https://github.com/helgeerbe/OpenDTU-OnBattery/assets/73672554/c4b9b06a-1ad5-4cce-bd38-7530cf011e2a">


 - egal wie ich die zwei GPIOs bzgl. TX/RX konfiguriere, in der Liveview wird zum MPPT nichts angezeigt. Muss ich auch noch GND verbinden? Bzw. kann man sehen, ob der ESP überhaupt etwas empfängt?

Der TX Port am MPPT ist auf "Normale Kommunikation" gestellt.

**Reply by spcqike** [reply]
Ja du brauchst ground ebenfalls.

**Reply by swingstate** [reply]
Und den führe ich dann auf einen der GND PINs des ESP, korrekt? Ich war bisher der Annahme, dass es mit 2 (TX/RX) Adern getan ist. Danke!

**Reply by frankh-gituser** [reply]
Bin auch gerade etwas ratlos. Habe mir einen 38-pin ESP32 als Testsystem bzw fuer den zweiten Victron mit der neuesten Release geflashed. Ein gleiches ESP Modell läuft mit der älteren openDTU Victron Release . Klemme ich das Victron Kabel vom alten zum neuen ESP bekommt der neue kein Signal. Hatte bisher Victron-pin-3 auf GPIO22 - habs jetzt auch mit GPIO32 versucht. Beim ersten gehts, beim zweiten nicht. Den Hoymiles erkennt diér neue ESP einwandfrei.

**Comment by swingstate** [comment]
#47 war die Lösung. 

Allerdings musste ich die Dose in welche der ESP geklebt war zur Hälfte zersägen um das PIN Mapping lesen zu können. Und siehe da,  GND war ganz woanders als vermutet, 6 PIN von unten gezählt. 

![IMG_1323](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/73672554/57d4a6bf-b973-499a-b024-80162ee53764)

Nun bin ich dabei die Ganzen Daten in Homeassistant einzubinden.

**Reply by madmartin** [reply]
Das hättest Du Dir sparen können, schliesslich hab ich Dir die Pin-Belegung der Module hier reingepostet....

**Reply by schlimmchen** [reply]
Den Gedanken hatte ich ehrlich gesagt auch :wink:

Und noch ein Hinweis: Wenn du das Gehäuse loswerden willst, dann erhitze doch das ganze Ding mit einem Heißluftfön, bis du das Board vom Gehäuse trennen kannst. Der Heißkleber wird recht schnell hinreichend "flüssig".

**Reply by swingstate** [reply]
ihr glaubt nicht wie oft ich mir die beiden Bilder angeschaut habe - aber den GND auf dem PIN habe ich nicht gesehen.

**Reply by schlimmchen** [reply]
Wir sind davon ausgegangen, dass du erkennst, dass es sich um einen generischen 38-pin NodeMCU handelt. Dann muss du das Bild, das Martin geteilt hat, im Kopf einmal umdrehen, und dann kannst du die Pins abzählen.

Schön jedenfalls, dass es jetzt funktioniert bei dir! Viel Erfolg noch!
